### PR TITLE
Fix `#[global] (Hint) Opaque` for prim. projections in sections

### LINKED ISSUE #19712: Anomaly when making projections hint opaque in a section in 8.20
### Description of the problem

git bisect points to:

```
commit 2d68d2d362d2a9433a20e16e048fa2656339f151
Author: Jan-Oliver Kaiser <janno@mpi-sws.org>
Date:   Wed Mar 13 16:01:14 2024 +0100

    Restore ability to make primitive projections hint opaque.
```

Oops.

### Small Rocq / Coq file to reproduce the bug

```coq
#[projections(primitive)]
Record test : Type := { f : nat }.
Section foo.
  #[global] Hint Opaque f : typeclass_instances.
End foo. (* Not_found *)
```

### Version of Rocq / Coq where this bug occurs

8.20

### Last version of Rocq / Coq where the bug did not occur

8.19

### PR DESCRIPTION
**Pull request by Janno**
Fixes / closes #19712 

- [x] Added / updated **test-suite**.

- [ ] Added **changelog**.

### CODE REVIEW

File: tactics/redexpr.ml [hunk 1]
```diff
@@ -99,7 +99,12 @@ let disch_ref ref =
   match ref with
   | Evaluable.EvalConstRef c -> Some ref
   | Evaluable.EvalProjectionRef p ->
-    let p = Lib.discharge_proj_repr p in
+    let p =
+      if Lib.is_in_section (Names.GlobRef.IndRef (Names.Projection.Repr.inductive p)) then
+        Lib.discharge_proj_repr p
```
Comment (SkySkimmer) [inline tactics/redexpr.ml:104]: possibly the change should be in `Lib.discharge_proj_repr` not in its callers
indeed coercion is one of the other users and has the same bug with
~~~
Set Primitive Projections.

Record test : Type := { f : nat }.
Section foo.
  Coercion f : test >-> nat.
End foo.
~~~

Comment (Janno) [reply tactics/redexpr.ml:104]: Oh, I should have checked. I saw this code and thought that there was one client that didn't need the check so I'd better not change `Lib.discharge_proj_repr`.

Comment (Janno) [reply tactics/redexpr.ml:104]: I changed the API. The original function still exists under a different name and could be exposed if necessary.

**Comment by Janno** [comment]
This probably deserves to be backported to 8.20 once it goes in.

File: tactics/hints.ml [hunk 2]
```diff
@@ -1302,7 +1302,12 @@ let discharge_autohint obj =
         let filter e = match e with
         | Evaluable.EvalConstRef c -> Some e
         | Evaluable.EvalProjectionRef p ->
-          let p = Lib.discharge_proj_repr p in
+          let p =
+            if Lib.is_in_section (Names.GlobRef.IndRef (Names.Projection.Repr.inductive p)) then
+              Lib.discharge_proj_repr p
+            else
+              p
+          in
```
Comment (silene) [inline tactics/hints.ml:1310]: I am confused. Is this code needed?

Comment (Janno) [reply tactics/hints.ml:1310]: Oh, I messed up my commit. Thanks for noticing!

**Review by SkySkimmer — APPROVED** [review]

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Comment by SkySkimmer** [comment]
@coqbot merge now

**Pull request merged**